docs/design/backup_and_restore.md

```diff
@@ -0,0 +1,202 @@
+* Desired state (running, stopped) (optional, defaults to running)
+* Key range to restore (optional, defaults to the full keyspace)
+* Version to restore (optional, defaults to latest version)
+* Timestamp to restore (optional, defaults to latest version)
+* Source cluster name (optional, required if using restore timestamps)
+* Restore tag (optional)
+
+All fields in a restore are immutable.
+
+If the restore version and restore timestamp are both specified, the version will take precedence.
+
+### Reconciling Backups
+
+The backup reconciler will run the following stages:
+
+* Check backup status
+* Configure backup agents
+* Toggle backup running
+* Modify backup
+* Toggle backup paused
+* Check backup status
+
+We will use the `fdbbackup` tool to modify the backup state. We will use the backup status that is found within the database status to determine the current state of the backup.
+
+#### Check Backup Status
+
+In the `Check Backup Status` stage, we will compare the running state and validate the following things:
+
+* The state of the backup matches the configured state in the spec
+* If the backup is configured to be running, it has completed an initial snapshot
+* The backup parameters match the parameters in the spec
+* The backup deployment exists, matches the deployment configuration for the spec, and is fully deployed
+
+If any of these checks fail, the backup will be marked as unreconciled, but reconciliation will continue through the later stages whether it is reconciled or not.
+
+#### Configure Backup Agents
+
+In the `Configure Backup Agents` stage, we will create a backup deployment that runs FDB's `backup_agent` process. The agents will be configured through a deployment object in the cluster spec. The deployment will have the following defaults:
+
+* The deployment will have the same name as the backup resource
+* The deployment will have 3 replicas
+* The deployment will have an init container running the `foundationdb/foundationdb-kubernetes-sidecar` image, configured to copy the cluster file from the same config map that the operator creates for the cluster.
+* The deployment will have a main container running the `foundationdb/foundationdb` image, running a `backup_agent` process.
+
+The version of FoundationDB that the backup agents run will be taken from the cluster resource. There must be a cluster resource in the same namespace with the name given in the cluster spec in order to run a backup. This version will also be used for the `fdbbackup` commands. In order to keep these versions in sync, the cluster reconciler will have a stage to check for backups that need to be updated. This stage will find any backup resources that are connected to the cluster and set an annotation, `foundationdb.org/backup-cluster-version` with the version in the cluster spec. During upgrades, the cluster reconciler will set this annotation, which will trigger a reconciliation of the backup that will update the backup agents.
```

Comment:
The last part implies that the cluster controller will change the deployment that is actually managed by the backup reconciler. Do you know if it's possible to watch for other resource events? I would prefer (if that's possible in a simple way) that the backup reconciler watched the cluster resources without the cluster reconciler having to interfere with the backup resources.

I think I spent some time in the past on that topic but couldn't find anything useful but maybe the controller runtime has some new features for that.

Reply:
The annotation I'm describing here would be set on the `FoundationDBBackup` resource, rather than the deployment. I don't know if there's a way for the backup reconciler to respond to events on the cluster object.